library/core/src/iter/traits/iterator.rs

```diff
@@ -2994,6 +2994,10 @@ pub trait Iterator {
     /// This is useful when you have an iterator over `&T`, but you need an
     /// iterator over `T`.
     ///
+    /// There is no guarantee whatsoever about the `clone` method actually
+    /// being called *or* optimized away. So code should not depend on
+    /// either.
+    ///
```

Comment:
This one seems a little more troublesome to me. I don't think this adds anything new, it simply makes the "no guarantee in either direction" which was previously implicit in the absence of detailed specification explicit. But some libs team members interpret API contracts differently than I do, so I want some extra eyes on this.

Reply:
At least making it explicit allows us to point the user to the docs when they bemoan their code breaking should we ever decide to improve perf on this.

Reply:
I don't mind making the lack of a guarantee explicit, I don't believe this restricts us in any way which is the main thing I worry about with docs changes